### PR TITLE
Default to JSON type if objects have an inconsistent structure

### PR DESCRIPTION
**Pull request by lnkuiper**
We always shredded the whole structure, but this has performance implications if the JSON is too inconsistent. Now, if a JSON objects' fields are too inconsistent, we default to the JSON type instead of a STRUCT.

### CODE REVIEW

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good.

While I think this approach sounds good I have two doubts about it:

* The metric might be hard to explain to users. I would say this needs to be well-documented and should also be a configurable option, rather than only an internal detail. 
* This might change behavior of SQL that could result in breaking scripts randomly depending on which JSON files are sampled.

For example, the following query will work fine while "field" is emitted as a struct, but will not work when "field" is emitted as JSON:

```sql
SELECT field.subfield FROM 'data/*.json';
```

That might not be as big of a deal as that would already break randomly with inconsistent schemas, though.

**Comment by lnkuiper** [comment]
Thanks for the review!

1. Agree that the metric is hard to explain. I will add it as a parameter and write some documentation about it.
2. This will always work once we implement the SQL:2023 standard which describes that JSON subfields should be extracted using 'dot'-notation. So even if "field" is a JSON column instead of a STRUCT, we should be able to rewrite the query to `json_extract` the "subfield" instead.

**Comment by lnkuiper** [comment]
Failure is unrelated, this is ready to go

File: extension/json/include/json.json [hunk 1]
```diff
@@ -116,22 +116,27 @@
       },
       {
         "id": 109,
+        "name": "field_appearance_threshold",
```
Comment (Mytherin) [inline extension/json/include/json.json:119]: This is a backwards incompatible change - the backwards compatible way of adding a new field is to add it to the end (with a previously unused field id) and to add a `default` value that is inline with what you expect if the field has not been set by the user (i.e. when reading an older version).

**Review by Mytherin — COMMENTED** [review]
Thanks! Looks good - one comment. I will retarget this to feature though as this is in my opinion too large of a change for a bug-fix release (and also isn't strictly a bug-fix).

**Comment by lnkuiper** [comment]
This was marked as high priority in our internal issues, so I though we wanted this in our 0.9.1 release. Should the branch be changed back to `main` again?

Also, thanks for clarifying the serialization, I didn't know how it worked. I hope I did it right now.

**Comment by Mytherin** [comment]
> This was marked as high priority in our internal issues, so I though we wanted this in our 0.9.1 release. Should the branch be changed back to main again?

The issue was marked as high priority due to who opened it, but I think this is more of a feature than a bug so let's leave it in the feature branch.

> Also, thanks for clarifying the serialization, I didn't know how it worked. I hope I did it right now.

Yes LGTM thanks

**Comment by dphugo** [comment]
Unfortunately, I've experienced the potential issue raised by @Mytherin.

> While I think this approach sounds good I have two doubts about it:
> 
> * The metric might be hard to explain to users. I would say this needs to be well-documented and should also be a configurable option, rather than only an internal detail.
> * This might change behavior of SQL that could result in breaking scripts randomly depending on which JSON files are sampled.

I have deeply nested and complex JSON that I need to parse and the unnest function worked brilliantly to interpret everything as STRUCT types. v0.8.1:
image

That was until I upgraded to v0.10.0, where the final and complex JSON is now parsed as "json" datatype:
image

It took me a few hours to find that it was this PR that caused this behavior and I agree it is confusing that it sometimes interprets the data as a STRUCT and other times as "json". I was unable to find any documentation on the "json" datatype. 
- Could you maybe point me to where the "json" datatype is explained and how to parse it from row data, please? 

A "configurable option" was mentioned, but I haven't found any documentation on this. 
- Could you please help me out how I can disable this feature via the configuration as mentioned above? I'm willing to take the speed/performance hit.

I would love to use the v0.10.1 as I have experienced the speed benefits, but have had to revert back to v0.8.1.

**Comment by lnkuiper** [comment]
@dphugo The JSON type is explained on the [JSON documentation page](https://duckdb.org/docs/extensions/json.html#json-type), like most other things related to JSON in DuckDB.

The old behavior is still possible in DuckDB using `field_appearance_threshold=0`.

This parameter has not been documented, but we are in the process of changing things as we also want to detect the `MAP` type.

**Pull request merged**